File: poc-csharp/RpsPoc/Concurrency/Tournament.cs
using System.Collections.Concurrent;

using RpsPoc.Domain;

namespace RpsPoc.Concurrency;

/// <summary>Dono do estado compartilhado e do ciclo de vida das tasks.</summary>
public sealed class Tournament
{
    private readonly TournamentConfig _config;
    private readonly EventBus _bus;
    private readonly ConcurrentDictionary<int, Match> _activeMatches = new();
    private readonly SemaphoreSlim _slots;

    private CancellationTokenSource? _cts;

    public Tournament(TournamentConfig config, EventBus bus)
    {
        _config = config;
        _bus = bus;
        _slots = new SemaphoreSlim(config.Threads, config.Threads);

        var players = new List<Player>(config.Players);
        for (var i = 1; i <= config.Players; i++)
        {
            players.Add(new Player(i));
        }

        Players = players;
        Queue = new PlayerQueue(players);
    }

    public PlayerQueue Queue { get; }

    public IReadOnlyList<Player> Players { get; }

    public IEnumerable<Match> ActiveMatches => _activeMatches.Values;

    public int ActiveCount => _activeMatches.Count;

    /// <summary>RF09.</summary>
    public bool IsFinished => Queue.Alive <= 1 && _activeMatches.IsEmpty;

    public Player? Champion => Queue.Champion();

    /// <summary>UC01. Retorna na hora: tudo acontece fora da UI thread (RNF01).</summary>
    public Task Start()
    {
        _cts = new CancellationTokenSource();
        _bus.Publish(new QueueChanged(Queue.Snapshot(), Queue.Alive));

        var orchestrator = new Orchestrator(Queue, _bus, _activeMatches, _slots, _config.RoundDelayMs);
        return Task.Run(() => orchestrator.RunAsync(_cts.Token), CancellationToken.None);
    }

    public void Stop()
    {
        Queue.Shutdown();
        _cts?.Cancel();
    }
}

File: src/main/java/rps/concurrency/Tournament.java
package rps.concurrency;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Semaphore;

import rps.domain.Match;
import rps.domain.Player;

/** Dono do estado compartilhado e do ciclo de vida das threads. */
public final class Tournament {

    private final TournamentConfig config;
    private final EventBus bus;
    private final PlayerQueue queue;
    private final List<Player> players;
    private final Set<Match> activeMatches = ConcurrentHashMap.newKeySet();

    private ExecutorService pool;
    private Thread orchestratorThread;

    public Tournament(TournamentConfig config, EventBus bus) {
        this.config = config;
        this.bus = bus;
        this.players = new ArrayList<>(config.players());
        for (int i = 1; i <= config.players(); i++) {
            players.add(new Player(i));
        }
        this.queue = new PlayerQueue(players);
    }

    /** UC01. Retorna imediatamente: tudo acontece fora da UI thread (RNF01). */
    public void start() {
        pool = Executors.newFixedThreadPool(config.threads(), runnable -> {
            Thread thread = new Thread(runnable, "match-worker");
            thread.setDaemon(true);
            return thread;
        });
        bus.publish(new Events.QueueChanged(queue.snapshot(), queue.alive()));

        // RF02: T vagas de partida simultanea. E o semaforo — nao o tamanho do
        // pool — que segura o orquestrador, mantendo na fila todo mundo que
        // ainda nao tem thread pra jogar. Em .NET o equivalente e o
        // SemaphoreSlim, entao as duas stacks passam a expressar T do mesmo jeito.
        Semaphore slots = new Semaphore(config.threads());

        // RF12: o relogio do torneio comeca na construcao do TournamentStats,
        // entao ele nasce aqui — o mais perto possivel da largada.
        TournamentStats stats = new TournamentStats();

        orchestratorThread = new Thread(
                new Orchestrator(queue, pool, bus, activeMatches, config.roundDelayMs(), slots, stats),
                "orchestrator");
        orchestratorThread.setDaemon(true);
        orchestratorThread.start();
    }

    public void stop() {
        if (orchestratorThread != null) {
            orchestratorThread.interrupt();
        }
        if (pool != null) {
            pool.shutdownNow();
        }
    }

    /** RF09. */
    public boolean isFinished() {
        return queue.alive() <= 1 && activeMatches.isEmpty();
    }

    public Player champion() {
        return queue.champion();
    }

    public PlayerQueue queue() {
        return queue;
    }

    public Set<Match> activeMatches() {
        return activeMatches;
    }

    public List<Player> players() {
        return players;
    }

    public TournamentConfig config() {
        return config;
    }
}
